packages/react-aria-components/src/Tree.tsx

```diff
@@ -194,11 +219,128 @@ function TreeInner<T extends object>({props, collection, treeRef: ref}: TreeInne
     layoutDelegate
   }, state, ref);
 
+  let dragState: DraggableCollectionState | undefined = undefined;
+  let dropState: DroppableCollectionState | undefined = undefined;
+  let droppableCollection: DroppableCollectionResult | undefined = undefined;
+  let isRootDropTarget = false;
+  let dragPreview: JSX.Element | null = null;
+  let preview = useRef<DragPreviewRenderer>(null);
+
+  if (hasDragHooks && dragAndDropHooks) {
+    dragState = dragAndDropHooks.useDraggableCollectionState!({
+      collection: state.collection,
+      selectionManager: state.selectionManager,
+      preview: dragAndDropHooks.renderDragPreview ? preview : undefined
+    });
+    dragAndDropHooks.useDraggableCollection!({}, dragState, ref);
+
+    let DragPreview = dragAndDropHooks.DragPreview!;
+    dragPreview = dragAndDropHooks.renderDragPreview
+      ? <DragPreview ref={preview}>{dragAndDropHooks.renderDragPreview}</DragPreview>
+      : null;
+  }
+
+  if (hasDropHooks && dragAndDropHooks) {
+    dropState = dragAndDropHooks.useDroppableCollectionState!({
+      collection: state.collection,
+      selectionManager: state.selectionManager
+    });
+    let dropTargetDelegate = dragAndDropHooks.dropTargetDelegate || ctxDropTargetDelegate || new dragAndDropHooks.ListDropTargetDelegate(state.collection, ref, {direction});
+    let keyboardDelegate = props.keyboardDelegate ||
+      new ListKeyboardDelegate({
+        collection: state.collection,
+        collator,
+        ref,
+        disabledKeys: state.selectionManager.disabledKeys,
+        disabledBehavior: state.selectionManager.disabledBehavior,
+        direction,
+        layoutDelegate
+      });
+    droppableCollection = dragAndDropHooks.useDroppableCollection!(
+      {
+        keyboardDelegate,
+        dropTargetDelegate,
+        onDropActivate: (e) => {
+          // Expand collapsed item when dragging over
+          if (e.target.type === 'item') {
+            let key = e.target.key;
+            let item = state.collection.getItem(key);
+            if (item && item.hasChildNodes && expandedKeys !== 'all' && !expandedKeys.has(key)) {
+              state.toggleKey(key);
+            }
+          }
+        },
+        onKeyDown: e => {
+          if (e.key === 'ArrowRight' || e.key === 'ArrowLeft') {
+            // Toggle expansion when drop target is after expandable item
+            let target = dropState?.target;
+            if (target && target.type === 'item') {
```

Comment:
We may want to update this behavior in the future and make it possible for the user to keyboard navigate onto the row when drag and dropping and then accept "ArrowRight/ArrowLeft". Feels a bit confusing at the moment